### PR TITLE
provision/kubernetes: add topologySpreadConstraint support

### PR DESCRIPTION
**Pull request by morpheu**
Add new cluster option "topology-spread-constraints" to specify topology constraints

### CODE REVIEW

File: provision/kubernetes/helpers.go [hunk 1]
```diff
@@ -1357,3 +1358,43 @@ func ignoreBaseDep(versionedGroup map[int][]deploymentInfo) {
...
+
+func topologySpreadConstraints(labels map[string]string, topologySpreadConstraintRule string) ([]apiv1.TopologySpreadConstraint, error) {
+	if topologySpreadConstraintRule == "" {
+		return nil, nil
+	}
+	var topologySpreadConstraintList []map[string]interface{}
+	err := json.Unmarshal([]byte(topologySpreadConstraintRule), &topologySpreadConstraintList)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse JSON object for topologySpreadConstraint: %v", err)
+	}
```
Comment (nettoclaudio) [inline provision/kubernetes/helpers.go:1370]: Why not just decode the JSON into a variable of type `[]apiv1.TopologySpreadConstraint`?

File: provision/kubernetes/helpers.go [hunk 2]
```diff
@@ -1357,3 +1358,43 @@ func ignoreBaseDep(versionedGroup map[int][]deploymentInfo) {
...
+		constraint := apiv1.TopologySpreadConstraint{
+			MaxSkew:     (int32)(maxSkew),
+			TopologyKey: topologyKey,
+			LabelSelector: &metav1.LabelSelector{
+				MatchLabels: filterAppLabels(labels),
+			},
+			WhenUnsatisfiable: apiv1.ScheduleAnyway,
+		}
+		topologySpreadConstraints = append(topologySpreadConstraints, constraint)
+	}
```
Comment (nettoclaudio) [inline provision/kubernetes/helpers.go:1387]: I think doesn't make sense forcing some default values here - such as WhenUnsatisfiable to "ScheduleAnyway". I'd keep only the inclusion of app's labels in the label selector instead.

```suggestion
	for _, c := range topologySpreadConstraintList {
		if c.LabelSelector == nil {
			c.LabelSelector = &metav1.LabelSelector{}
		}

		if c.LabelSelector.MatchLabels == nil {
			c.LabelSelector.MatchLabels = make(map[string]string)
		}

		for k, v := range filterAppLabels(labels) {
			c.LabelSelector.MatchLabels[k] = v
		}
		
		topologySpreadConstraints = append(topologySpreadConstraints, c)
	}
```

File: provision/kubernetes/helpers.go [hunk 3]
```diff
@@ -1357,3 +1358,43 @@ func ignoreBaseDep(versionedGroup map[int][]deploymentInfo) {
...
+
+func filterAppLabels(labels map[string]string) map[string]string {
+	searchLabels := []string{tsuruLabelAppName, tsuruLabelAppProcess, tsuruLabelAppVersion}
+	filteredLabels := make(map[string]string)
+	for _, label := range searchLabels {
+		if value, ok := labels[label]; ok {
+			filteredLabels[label] = value
+		}
+	}
+	return filteredLabels
```
Comment (nettoclaudio) [inline provision/kubernetes/helpers.go:1399]: Not a big thing, just removing unecessary iteration over app labels.

```suggestion
	filtered := map[string]string{
		tsuruLabelAppName:    labels[tsuruLabelAppName],
		tsuruLabelAppProcess: labels[tsuruLabelAppProcess],
	}

	if v, found := labels[tsuruLabelAppVersion]; found {
		filtered[tsuruLabelAppVersion] = v
	}

	return filtered
```

**Review by nettoclaudio — CHANGES_REQUESTED** [review]

**Review by nettoclaudio — APPROVED** [review]
lgtm

**Review by wpjunior — APPROVED** [review]

**Pull request merged**